addons/binding/org.openhab.binding.hs110/src/main/java/org/openhab/binding/hs110/discovery/HS110DiscoveryService.java

```diff
@@ -0,0 +1,156 @@
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.SocketTimeoutException;
+import java.net.UnknownHostException;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.hs110.HS110BindingConstants;
+import org.openhab.binding.hs110.internal.HS110;
+import org.openhab.binding.hs110.internal.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
+ * The {@link HS110DiscoveryService2} detects new Power Plugs by sending a UDP network broadcast
+ * and parsing the answer into a thing.
+ *
+ * @author Christian Fischer - Initial contribution
+ */
+public class HS110DiscoveryService extends AbstractDiscoveryService {
+
+    private Logger logger = LoggerFactory.getLogger(HS110DiscoveryService.class);
+
+    ///// Network
+    private byte[] discoverbuffer = Util.encryptBytes(HS110.Command.SYSINFO.value);
+    private final DatagramPacket discoverPacket;
+    private byte[] buffer = new byte[1024];
+    private DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
+
+    private final InetAddress broadcast;
+
+    private Future<?> scanFuture;
+    private static final int DISCOVER_TIMEOUT_SECONDS = 5;
+    private static final int RECEIVE_JOB_TIMEOUT = 200;
+    private static final int UDP_PACKET_TIMEOUT = RECEIVE_JOB_TIMEOUT - 50;
+
+    public HS110DiscoveryService() throws UnknownHostException {
+
+        super(HS110BindingConstants.SUPPORTED_THING_TYPES_UIDS, DISCOVER_TIMEOUT_SECONDS, false);
+        byte[] addr = { (byte) 0xff, (byte) 0xff, (byte) 0xff, (byte) 0xff };
+        broadcast = InetAddress.getByAddress(addr);
+        discoverPacket = new DatagramPacket(discoverbuffer, discoverbuffer.length, broadcast, 9999);
+
+    }
+
+    @Override
+    protected void startScan() {
+        logger.debug("Start Scan for {}",
+                this.getSupportedThingTypes().stream().map(uid -> uid.toString()).collect(Collectors.joining(" ")));
+        discoverThings();
+
+    }
+
+    @Override
+    protected void stopScan() {
+        logger.debug("Stop Scan for {}",
+                this.getSupportedThingTypes().stream().map(uid -> uid.toString()).collect(Collectors.joining(" ")));
+
+        if (scanFuture != null) {
+            scanFuture.cancel(true);
+        }
+        super.stopScan();
+    }
+
+    private DatagramSocket datagramSocket;
+
+    private void discoverThings() {
+        if (scanFuture == null) {
+            try {
+                datagramSocket = new DatagramSocket(null);
```

Comment:
You should close this socket, maybe use the [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) then it will be properly closes automagicly